Firestore/core/test/firebase/firestore/model/field_value_test.cc

```diff
@@ -0,0 +1,169 @@
+
+#include <vector>
+
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+using Type = FieldValue::Type;
+
+TEST(FieldValue, NullType) {
+  const FieldValue value = FieldValue::NullValue();
+  EXPECT_EQ(Type::Null, value.type());
+  EXPECT_FALSE(value < value);
+}
+
+TEST(FieldValue, BooleanType) {
+  const FieldValue true_value = FieldValue::BooleanValue(true);
+  const FieldValue false_value = FieldValue::BooleanValue(false);
+  EXPECT_EQ(Type::Boolean, true_value.type());
+  EXPECT_FALSE(true_value < true_value);
+  EXPECT_FALSE(true_value < false_value);
+  EXPECT_FALSE(false_value < false_value);
+  EXPECT_TRUE(false_value < true_value);
+}
+
+TEST(FieldValue, ArrayType) {
+  const FieldValue empty =
+      FieldValue::ArrayValue(std::vector<const FieldValue>{});
+  std::vector<const FieldValue> small_value_array = {
+      FieldValue::NullValue(), FieldValue::BooleanValue(true),
+      FieldValue::BooleanValue(false)};
+  // copy small array
+  const FieldValue small = FieldValue::ArrayValue(small_value_array);
+  std::vector<const FieldValue> large_value_array = {
+      FieldValue::BooleanValue(true), FieldValue::BooleanValue(false)};
+  // move large array
+  const FieldValue large =
+      std::move(FieldValue::ArrayValue(std::move(large_value_array)));
+  EXPECT_EQ(Type::Array, empty.type());
+  EXPECT_EQ(Type::Array, small.type());
+  EXPECT_EQ(Type::Array, large.type());
+  EXPECT_TRUE(empty < small);
+  EXPECT_FALSE(small < empty);
+  EXPECT_FALSE(small < small);
+  EXPECT_TRUE(small < large);
+  EXPECT_FALSE(large < small);
+}
+
+TEST(FieldValue, Assignment) {
+  FieldValue clone = FieldValue::TrueValue();
+  const FieldValue null_value = FieldValue::NullValue();
+  clone = null_value;
+  EXPECT_EQ(FieldValue::NullValue(), clone);
+  clone = clone;
+  EXPECT_EQ(FieldValue::NullValue(), clone);
+
+  const FieldValue true_value = FieldValue::TrueValue();
+  clone = true_value;
+  EXPECT_EQ(FieldValue::TrueValue(), clone);
+  clone = clone;
+  EXPECT_EQ(FieldValue::TrueValue(), clone);
+  clone = null_value;
+  EXPECT_EQ(FieldValue::NullValue(), clone);
+
+  const FieldValue array_value =
+      FieldValue::ArrayValue(std::vector<const FieldValue>{
+          FieldValue::TrueValue(), FieldValue::FalseValue()});
+  clone = array_value;
+  EXPECT_EQ(FieldValue::ArrayValue(std::vector<const FieldValue>{
+                FieldValue::TrueValue(), FieldValue::FalseValue()}),
+            clone);
+  clone = clone;
+  EXPECT_EQ(FieldValue::ArrayValue(std::vector<const FieldValue>{
+                FieldValue::TrueValue(), FieldValue::FalseValue()}),
+            clone);
+  clone = null_value;
+  EXPECT_EQ(FieldValue::NullValue(), clone);
+}
+
+TEST(FieldValue, Move) {
+  FieldValue clone = FieldValue::TrueValue();
+
+  FieldValue null_value = FieldValue::NullValue();
+  clone = std::move(null_value);
+  EXPECT_EQ(FieldValue::NullValue(), clone);
+  clone = std::move(clone);
```

Comment:
I'm really surprised this works.

The point of `std::move` to indicate that after the call that name will not be used again. The value with that name becomes invalid immediately after `std::move` returns (but before the assignment).

See here for more discussion:
https://stackoverflow.com/questions/24604285/behaviour-of-move-assignment-to-self

Sorry, I should have been clearer: it's the copy constructor/copy assignment that need to work when copying *this.

Reply:
This works since the move operator= implementation for FieldValue uses std::swap and thus both sides are still well-defined after the call. I can go the other design (self-check+move), if that is more preferred. How do you think?